### ISSUE TITLE
TimestampWriter isn't being used

### ISSUE THREAD

**Issue by ashleycoxley** [apache/incubator-iceberg#475]
When trying to write to a field with type `Types.TimestampType.withoutZone()`, I'm getting the exception 
`java.time.LocalDateTime cannot be cast to java.time.OffsetDateTime`. This appears to be due to a case match miss in the `GenericParquetWriter` - for all primitives of type `TIMESTAMP_MICROS`, the `TimestamptzWriter` is returned. I believe it should only be returned when `shouldAdjustToUTC` is true; and the `TimestampWriter` returned otherwise  (similar to the implementation in `GenericParquetReaders`).

**Comment by rdblue** [comment]
You're right. The reason why this is a problem is that Parquet 1.10.1 doesn't have the `shouldAdjustToUTC` flag, and `GenericParquetWriter` uses only the Parquet schema to construct the writer. So there's no way to know whether the timestamp is with time zone or without time zone.

I think we could fix this by also passing in the Iceberg schema, which has the correct type. I think this would be fairly easy if we converted the builder to use `TypeWithSchemaVisitor`, which traverses a Parquet schema and an Iceberg schema at the same time.

**Comment by wangmiao1981** [comment]
@rdblue Just hit the same issue. In TypeToMessageType.java, 
https://github.com/apache/incubator-iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/TypeToMessageType.java#L82, calls https://github.com/apache/incubator-iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/TypeToMessageType.java#L133
It can't store the `shouldAdjustToUTC ` field.  This is the reason it can't call the correct writer. So it fails in the `if` branch, https://github.com/apache/incubator-iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java#L212

There are few other existing tests pass, which go to https://github.com/apache/incubator-iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java#L235, which ignores the boolean flag and write it as Timestamp.withZone(). So it passes. 

In Spark parquet writer, it actually throws an exception:

`java.lang.UnsupportedOperationException: Spark does not support timestamp without time zone field`

If I understand correctly, it may not be fixed by using `TypeWithSchemaVisitor ` only. I will give it a try. I have made changes to throw similar exception as Spark does and comment out some tests. That was done before I found this opened issue.

If it is not an easy fix, I will submit the exception PR.

I also did a quick test on UUID and it also fails. I haven't debug that issue yet. I suspect similar issue of Logical Type support.

**Comment by rdblue** [comment]
I just took a look and it wasn't too bad to fix it by adding support for Parquet's new logical type annotations, which came out in 1.11.0 and store `isAdjustedToUTC`. Now, places that assumed the type was `timestamptz` can correctly distinguish between the two timestamp types. The PR is #891. We may need to add support in a few more places, I didn't have time to look very thoroughly.

**Comment by wangmiao1981** [comment]
+1 to add logical type. Will take a look the PR above. Just started reading source code. I may give a try. 

Thanks!
